### PR TITLE
CX on Blockchain (CX chains) prototype

### LINKED ISSUE #309: Serialization is not taking into account non-default stack sizes.
**Describe the bug**
A CX program is not recording meta-data about its stack size. This is problematic for serialization because we can't know where the data segment starts. This error was not being detected because CX programs had a constant stack-size before (it was not configurable).

**To Reproduce**
1. Deserialize a CX program that was serialized with a stack size different than the stack size set in the CX installation that is deserializing the program.

**Expected behavior**
A CX program needs to record its stack size.

**Desktop (please complete the following information):**
 - CX Version [e.g. 0.6.2]

### LINKED ISSUE #310: Splitting a serialized program into its blockchain and transaction parts
**Describe the bug**
We need a way to split a serialized program in order to separate the blockchain code (the serialization with the program state) and the transaction code (the serialization containing the instructions to be run with the stored program state).

**To Reproduce**
1. Run a CX program in `--transaction` mode to try a transaction program using the program state stored on the blockchain.

**Expected behavior**
A serialized CX program needs to be "splittable".

**Desktop (please complete the following information):**
 - CX Version [e.g. 0.6.2]

### LINKED ISSUE #311: `CurrentFunction` and `CurrentStruct` are causing errors in programs with more than 1 package
**Describe the bug**
The indexes used to refer to serialized struct and functions in packages do not match the indexes used in `CurrentFunction` and `CurrentStruct` fields of the package. For example, if there are 2 packages, one with 100 functions and another with 1, the `CurrentFunction` index of the second package is going to be 100, when it had to be 0 (the first and only one).

**To Reproduce**
1. Serialize and deserialize a CX program with a number of packages greater than 1, and each of these packages must have at least 1 function and/or struct definition.

**Expected behavior**
The indexes for the `CurrentFunction` and `CurrentStruct` of each serialized package must correspond to the indexes set in the deserialized program.

**Desktop (please complete the following information):**
 - CX Version [e.g. 0.6.2]

### LINKED ISSUE #312: Deserialization is not setting correctly the sizes for the CallStack, HeapStartsAt and StackSize fields of the CXProgram structure
**Describe the bug**
The process of deserialization is not considering variable sizes for the `CallStack`, `HeapStartsAt`
 and `StackSize` fields of the `CXProgram` structure. Deserializing a CX program using a CX installation that is configured to use different values for the stack and heap size will cause a crash.

**To Reproduce**
1. Deserialize a CX program that was serialized with a stack size, heap size or callstack size different than the ones set in the CX installation that is deserializing the program.

**Expected behavior**
Deserializing a CX program needs to consider variable stack and heap sizes.

**Desktop (please complete the following information):**
 - CX Version [e.g. 0.6.2]

### PR DESCRIPTION
**Pull request by amherag**
Fixes:
* fixes #309, fixes #310, fixes #311, fixes #312

Changes:
* Vendored Skycoin
* Initial code to run a CX program on the Skycoin blockchain
* Fixed several issues with serialization and deserialization

Does this change need to mentioned in CHANGELOG.md?
Yes.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,14 +2,20 @@
 
 ### v0.6.3 (NOT YET RELEASED)
 * Additions
-  * ...
+  * Prototype for CX chains (CX programs stored on the Skycoin blockchain)
 * Libraries
 * Fixed issues
-  * #xx: ...
+  * #309: Serialization is not taking into account non-default stack sizes.
```
Comment (ingwal) [inline CHANGELOG.md]: All of these fixes should be in their own PR unless they are fixed by the same code.

File: cx/config.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package base
+package cx
```
Comment (ingwal) [inline cx/config.go]: I don't think this is a good choice of package name.  "cx" is probably a better name than "base" for packages outside of CX that want to include it.  After all, "base" could be anything.

On the other hand, we still want to keep the 3(?) levels of CX packages where base is only the first level (I don't remember the names of the other 2 right now).  So I suggest that we instead use "cxbase" for the base CX level and use the "cx" prefix also for the other levels.

File: cx/cxprogram.go [hunk 3]
```diff
@@ -33,6 +33,7 @@ type CXProgram struct {
 	Inputs       []*CXArgument
 	Outputs      []*CXArgument
 	Memory       []byte // Used when running the program
+	StackSize    int
```
Comment (ingwal) [inline cx/cxprogram.go]: Maybe add a little documentation to this field? :)

(and the others too...)

File: cx/op_aff.go [hunk 4]
```diff
@@ -619,15 +619,15 @@ func opAffOf(expr *CXExpression, fp int) {
...
+	// prevStrct := prevPkg.CurrentStruct
 	prevFn := prevPkg.CurrentFunction
 	prevExpr := prevFn.CurrentExpression
 
 	var tgtPkg = CXPackage(*expr.Package)
-	var tgtStrct CXStruct
-	if prevStrct != nil {
-		tgtStrct = *prevStrct
-	}
+	// var tgtStrct CXStruct
```
Comment (ingwal) [inline cx/op_aff.go]: I think that you should either remove old code like this if it will never be needed again or comment on why it's disabled.  As it is now, it will probably lay around forever and nobody will remember why it was there and/or why it isn't anymore.

File: cxgo/cxgo.nex [hunk 5]
```diff
@@ -206,14 +209,23 @@ import (
...
+	"github.com/skycoin/skycoin/src/readable"
+	"github.com/skycoin/skycoin/src/skycoin"
+	"github.com/skycoin/skycoin/src/util/logging"
+	"github.com/skycoin/skycoin/src/visor"
+
+	"errors"
 )
 
-const VERSION = "0.6.1"
+const VERSION = "0.6.2"
```
Comment (ingwal) [inline cxgo/cxgo.nex]: This fix should really be in 0.6.2 already.  We need to add changing the version number into the release todo file.

File: cxgo/cxgo.nex [hunk 6]
```diff
@@ -613,6 +625,264 @@ func parseMemoryString (s string) int {
...
+		if err != nil {
+			log.Error("Error with NewVisor")
+			return err
+		}
+
+		headSeq, _, err := vs.HeadBkSeq()
+		if err != nil {
+			log.Error("Error with HeadBkSeq")
+			return err
+		} else if headSeq == 0 {
```
Comment (ingwal) [inline cxgo/cxgo.nex]: You don't need the 'else' here since the err != nil part above has a return in it.

File: cxgo/cxgo.nex [hunk 7]
```diff
@@ -613,6 +625,264 @@ func parseMemoryString (s string) int {
...
+		}
+	} else {
+		log.Error("error getting genesis block")
+	}
+	return err
+}
+
+// end
+
+func runNode (mode string, options cxCmdFlags) *exec.Cmd {
```
Comment (ingwal) [inline cxgo/cxgo.nex]: Wouldn't 'mode' fit better as an enum?

File: cxgo/cxgo.nex [hunk 8]
```diff
@@ -625,6 +895,39 @@ func main () {
 	registerFlags(&options)
 	flag.Parse()
 
+	if options.publisherMode || options.peerMode {
```
Comment (ingwal) [inline cxgo/cxgo.nex]: Unless there can be >1 mode active at the same time, I suggest using a value to represent the mode (int?) instead of a set of bools.

File: cxgo/cxgo.nex [hunk 9]
```diff
@@ -1024,22 +1349,228 @@ func main () {
...
-		if AssertFailed() {
-			os.Exit(CX_ASSERT)
+
+		if options.blockchainMode {
+			PRGRM.RemovePackage(MAIN_FUNC)
+			
+			s := Serialize(PRGRM, 1)
+
+			
+
```
Comment (ingwal) [inline cxgo/cxgo.nex]: Huh?  Why all the empty lines?

File: cxgo/flags.go [hunk 10]
```diff
@@ -22,6 +22,19 @@ type cxCmdFlags struct {
 	initialHeap         string
 	maxHeap             string
 	stackSize           string
+	blockchainMode      bool
```
Comment (ingwal) [inline cxgo/flags.go]: See comment above...  Use an enum instead of n bools?

File: tests/main.cx [hunk 11]
```diff
@@ -437,6 +437,12 @@ func main ()() {
 	runTestEx("issue-294-a.cx", cx.SUCCESS, "Crash in garbage collector when heap is resized", TEST_ISSUE, 0)
 	runTestEx("issue-294-b.cx", cx.SUCCESS, "Crash in garbage collector when heap is resized", TEST_ISSUE, 0)
 
+	// We need to fix serialization and deserialization as user-callable functions
+	// runTestEx("issue-309.cx", cx.SUCCESS, "Serialization is not taking into account non-default stack sizes.", TEST_ISSUE, 0)
```
Comment (ingwal) [inline tests/main.cx]: Hmm, these are fixed now, aren't they?

File: cx/op_aff.go (hunk 4 above)
Comment (amherag) [reply cx/op_aff.go]: This is not old code. It will be used once we complete affordances. The linter was complaining because it was not being used and I had to temporarily remove it.

File: cx/cxprogram.go (hunk 3 above)
Comment (amherag) [reply cx/cxprogram.go]: Okay.

File: cx/config.go (hunk 2 above)
Comment (amherag) [reply cx/config.go]: "cxbase" sounds good. Or how about "cxlang"?

File: CHANGELOG.md (hunk 1 above)
Comment (amherag) [reply CHANGELOG.md]: I ended up with these fixes in this PR because they are mostly fixes to bugs that were brought by implementing the new features. So, for those bug fixes to "make sense" we'd need to first merge the features PR and then merge the bug fixes. This couldn't be avoided because in order to continue implementing the features I had to fix the bugs or create a PR with the unfinished and buggy features to later create the PRs with the bug fixes.

Anyway, now that we have everything we can split them and merge them in order.

File: cxgo/cxgo.nex (hunk 5 above)
Comment (amherag) [reply cxgo/cxgo.nex]: It was changed already, as far as I remember. I don't know if we added a PR where the version was set back to 0.6.1. But yeah, let's just be extra careful.

File: cxgo/cxgo.nex (hunk 6 above)
Comment (amherag) [reply cxgo/cxgo.nex]: Okay, I'll change it.

File: cxgo/cxgo.nex (hunk 7 above)
Comment (amherag) [reply cxgo/cxgo.nex]: Sure. I decided to use just a string because there are only two values and efficiency is not a concern in this case. I'll add the constants.

File: cxgo/cxgo.nex (hunk 8 above)
Comment (amherag) [reply cxgo/cxgo.nex]: Okay.

File: cxgo/cxgo.nex (hunk 9 above)
Comment (amherag) [reply cxgo/cxgo.nex]: It's work in progress. I'll remove them.

File: cxgo/flags.go (hunk 10 above)
Comment (amherag) [reply cxgo/flags.go]: These are used for the flag package. I could force a workaround to include constants there, but I guess that someone looking at that code is going to expect a field for each of the flags.

File: tests/main.cx (hunk 11 above)
Comment (amherag) [reply tests/main.cx]: They do not exist. We need to add `serialize` and `deserialize` as user-callable functions and then add those tests. The Github issues exist now.

File: CHANGELOG.md (hunk 1 above)
Comment (ingwal) [reply CHANGELOG.md]: Aha.  Well, in that case, I would be fine with even not mentioning them at all.  Let's discuss and come up with the best solution.

File: cx/config.go (hunk 2 above)
Comment (ingwal) [reply cx/config.go]: Well, we should come up with a consistent scheme for all our packages within CX.

File: cx/op_aff.go (hunk 4 above)
Comment (ingwal) [reply cx/op_aff.go]: Hmm...  Let's mark them clearly then with a comment and a specific tag so that it's easy to find again later.

File: tests/main.cx (hunk 11 above)
Comment (ingwal) [reply tests/main.cx]: Ok, I see.

File: CHANGELOG.md (hunk 1 above)
Comment (amherag) [reply CHANGELOG.md]: Removed them from the Changelog.

File: cx/config.go (hunk 2 above)
Comment (amherag) [reply cx/config.go]: It's cxcore now :)

File: cx/op_aff.go (hunk 4 above)
Comment (amherag) [reply cx/op_aff.go]: Didn't see this on time. Already removed :\ once I need them it should be easy to get the code from a previous git commit though.

File: cx/constcodes.go [hunk 12]
```diff
@@ -248,6 +250,15 @@ const (
 	CONST_CX_RUNTIME_SLICE_INDEX_OUT_OF_RANGE
 )
 
+const (
```
Comment (ingwal) [inline cx/constcodes.go]: I know that you don't name the enums in Golang, but can we have a comment saying what these constants are for?

File: cx/serialize.go [hunk 13]
```diff
@@ -637,82 +639,75 @@ func initSerialization(prgrm *CXProgram, s *sAll) {
...
 }
 
-// Serialize ...
-func Serialize(prgrm *CXProgram) (byts []byte) {
-	// prgrm.PrintProgram()
-
-	s := sAll{}
-	initSerialization(prgrm, &s)
-
+// SplitSerialize ...
```
Comment (ingwal) [inline cx/serialize.go]: At this point, I really think we should document our functions when we create new ones.

File: cx/serialize.go [hunk 14]
```diff
@@ -739,21 +734,34 @@ func Serialize(prgrm *CXProgram) (byts []byte) {
...
+					sFn.ExpressionsOffset, sFn.ExpressionsSize = serializeIntegers(exprs, s)
 				}
 			} else {
 				panic("function reference not found")
 			}
 		}
 	}
+}
+
+// Serialize ...
```
Comment (ingwal) [inline cx/serialize.go]: Again, I think the code is much easier to understand if we have the main functions near the top of the file.

File: cx/serialize.go [hunk 15]
```diff
@@ -843,6 +851,8 @@ func dsName(off int32, size int32, s *sAll) string {
 }
 
 func dsPackages(s *sAll, prgrm *CXProgram) {
+	var fnCounter int32
```
Comment (ingwal) [inline cx/serialize.go]: I think this may be another case of what I wrote about in an earlier PR: The name is misleading since it seems to store a size of some kind.  At least you add the value to a memory pointer later.

File: cx/op.go [hunk 16]
```diff
@@ -196,6 +196,10 @@ func MarkAndCompact() {
 	for c := 0; c <= PROGRAM.CallCounter; c++ {
 		op := PROGRAM.CallStack[c].Operator
 
+		if op == nil {
```
Comment (ingwal) [inline cx/op.go]: What does it mean when op == nil?  Comment, please.

File: cx/op_os.go [hunk 17]
```diff
@@ -16,15 +16,6 @@ import (
 	"github.com/skycoin/skycoin/src/cipher/encoder"
 )
 
-const (
```
Comment (ingwal) [inline cx/op_os.go]: Why the move of these constants?  To me it makes sense to have them in the OS releated file op_os.go...

File: cx/serialize.go [hunk 18]
```diff
@@ -1168,7 +1184,268 @@ func Deserialize(byts []byte) (prgrm *CXProgram) {
...
+	prgrmState = append(prgrmState, make([]byte, prgrm2Info.StackSize)...)
+	// We are only interested on extracting the data segment
+	prgrmState = append(prgrmState, (*sPrgrm1)[index1.NamesOffset+prgrm1Info.StackSize:index1.NamesOffset+prgrm1Info.StackSize+(prgrm2Info.HeapStartsAt-prgrm2Info.StackSize)]...)
+
+	for i, byt := range prgrmState {
+		(*sPrgrm2)[i+int(index2.MemoryOffset)] = byt
+	}
+}
+
+func correctSerializedSize(byts *[]byte, off1, off2 int32, n int) {
```
Comment (ingwal) [inline cx/serialize.go]: Document the function.

File: cx/serialize.go [hunk 19]
```diff
@@ -1168,7 +1184,268 @@ func Deserialize(byts []byte) (prgrm *CXProgram) {
...
+	if len((*byts)[off1:off2]) == 0 {
+		return
+	}
+	for i, byt := range encoder.SerializeAtomic(int32((off2 - off1 - 4) / int32(n))) {
+		(*byts)[off1+int32(i)] = byt
+	}
+}
+
+func mustSize(obj interface{}) int {
```
Comment (ingwal) [inline cx/serialize.go]: Document the function.  Also, the name: mustSize() does not give me any hint to what it does.

File: cx/op_os.go (hunk 17 above)
Comment (amherag) [reply cx/op_os.go]: If I move them back to op_os.go the linter goes crazy. To fix it I'd need to remove the build tags, and then a bunch of errors pops up. We'll be moving them back to op_os later.

File: cx/op.go (hunk 16 above)
Comment (amherag) [reply cx/op.go]: There are situations where calls are pushed to the call stack and they are "operatorless". As far as I remember it's something related to OpenGL (one of its functions that use callbacks).

File: cx/serialize.go (hunk 19 above)
Comment (amherag) [reply cx/serialize.go]: I grabbed the idea from the Skycoin code. Actually, I think "must" is better than "safe", as "safe" implies it's always going to return something without error. Something like a safe div, when a division by zero is always prevented. "must" on the other hand implies that it must return something or it will always stop. I'm open to other suggestions though.

File: cx/serialize.go (hunk 18 above)
Comment (amherag) [reply cx/serialize.go]: Done!
